module/observations/helper.js

```diff
@@ -2819,36 +2823,66 @@ module.exports = class ObservationsHelper {
 
 
-          let roles = bodyData.role.split(",");
+          let tenantDetails = await userService.fetchPublicTenantDetails(tenantData.tenantId);
+          if (!tenantDetails.success || !tenantDetails?.data?.meta) {
+            throw { 
+              status: httpStatusCode.bad_request.status,
+              message: messageConstants.apiResponses.FAILED_TO_FETCH_TENANT_DETAILS,
+            };
+          }
 
-          let entityTypeArr = [];
-          for(let roleIndex=0;roleIndex<roles.length;roleIndex++){
+          let observableEntityKeys = tenantDetails.data.meta?.observableEntityKeys || [];
 
-          let rolesDocumentAPICall = await entityManagementService.userRoleExtension({
-            code: roles[roleIndex],
-            tenantId:solutionDocument[0].tenantId,
-            orgIds:{$in:['ALL',...solutionDocument[0].orgIds]}
-          },
-          ["entityTypes.entityType"])
-
-          if(rolesDocumentAPICall.success){
-            entityTypeArr.push(rolesDocumentAPICall.data[0].entityTypes[0].entityType)
-          }else{
+          if (!observableEntityKeys || observableEntityKeys.length === 0) {
             throw {
               status: httpStatusCode.bad_request.status,
-              message: messageConstants.apiResponses.USER_ROLES_NOT_FOUND,
+              message: messageConstants.apiResponses.OBSERVABLE_ENTITY_KEYS_NOT_FOUND,
             };
           }
 
-        }
+          let KeytoValidate = [];
+
+          for (let i = 0; i < observableEntityKeys.length; i++) {
+            KeytoValidate.push(observableEntityKeys[i]);
+          }
+
+          let roles = observableEntityKeys
+            .filter((key) => typeof bodyData[key] === 'string' && bodyData[key].trim() !== '')
+            .flatMap((key) => bodyData[key].split(',').map((role) => role.trim()));
+
+          let entityTypeArr = [];
+
+          for (let roleIndex = 0; roleIndex < roles.length; roleIndex++) {
+            let rolesDocumentAPICall = await entityManagementService.entityDocuments(
+              {
+                _id: roles[roleIndex],
+                tenantId: solutionDocument[0].tenantId,
+                orgIds: { $in: ['ALL', solutionDocument[0].orgId] },
+              },
+              ['metaInformation.targetedEntityTypes']
+            );
+
+            if (
+              rolesDocumentAPICall?.success &&
+              Array.isArray(rolesDocumentAPICall.data) &&
+              rolesDocumentAPICall.data[0]?.metaInformation?.targetedEntityTypes?.[0]?.entityType
+            ) {
+              entityTypeArr.push(rolesDocumentAPICall.data[0].metaInformation.targetedEntityTypes[0].entityType);
+            } else {
+              throw {
+                status: httpStatusCode.bad_request.status,
+                message: messageConstants.apiResponses.USER_ROLES_NOT_FOUND,
+              };
+            }
+          }
 
           const uniqueEntityTypeArr = _.uniq(entityTypeArr);
```

Comment:
@borkarsaish65 The logic should be this variable array list should match the solution entity type, if yes, then allow to create, else don't allow

Reply:
done